
#' Format Whippet junctions as a GRanges object
#' @param junctions data.frame containing junction location information. May be generated by readWhippetJNCfiles()
#' @return GRanges object with junctions
#' @export
#' @import GenomicRanges
#' @examples
#' @author Beth Signal
formatJunctions <- function(junctions){
    jncRanges <- GRanges(seqnames=Rle(junctions$chrom),
                          ranges=IRanges(start=as.numeric(junctions$start),
                                         end=as.numeric(junctions$end)),
                          strand=junctions$strand, id=junctions$id)
    return(jncRanges)
}


#' Find alternative junctions for Whippet alternative splicing events
#'
#' Find junctions that pair with each end of an AA (alt. acceptor) or AD (alt. donor) whippet range
#' Find junctions that pair with the upsteam/downstream exon of an AF (alt. first exon) or an AL (alt. last exon)
#' @param exonRanges GRanges object with Whippet event coordinates
#' @param jncRanges GRanges object with Whippet junctions. Generated by formatJunctions()
#' @param type type of Whippet event (AA/AD/AF/AL). Note only one event type should be processed at a time.
#' @return GRanges object with alternative junctions. Each event should have a set of X (for which the psi measurement is reported) junctions, and alternative Y junctions.
#' @export
#' @import GenomicRanges
#' @examples
#' @author Beth Signal
findJunctionPairs <- function(exonRanges, jncRanges, type=NA){

    exonRanges$type <- type
    # search for alternatives to the left or the right?
    exonRanges$search <- "right"
    exonRanges$search[(exonRanges$type=="AA" & as.logical(strand(exonRanges) == '+'))|
                          (exonRanges$type=="AD" & as.logical(strand(exonRanges) == '-'))|
                          (exonRanges$type=="AF" & as.logical(strand(exonRanges) == '-'))|
                          (exonRanges$type=="AL" & as.logical(strand(exonRanges) == '+'))] <- "left"

    junction_SJA.right <- exonRanges[exonRanges$search=="right"]
    junction_SJA.left <- exonRanges[exonRanges$search=="left"]

    # right if AA&-
    # left if AA&+
    ol_A_from <- vector()
    if(length(junction_SJA.right) > 0){
        if(type %in% c("AA", "AD")){
            start(junction_SJA.right) <- start(junction_SJA.right) -1
            end(junction_SJA.right) <- start(junction_SJA.right)
        }else{
            start(junction_SJA.right) <- end(junction_SJA.right)
        }
        ol_A.right <- findOverlaps(junction_SJA.right, jncRanges, type="start")
        ol_A_from <- append(ol_A_from,
                            as.character(junction_SJA.right$id[ol_A.right@from]))
        junctions_A <- jncRanges[ol_A.right@to]
    }
    if(length(junction_SJA.left) > 0){
        end(junction_SJA.left) <- start(junction_SJA.left)
        ol_A.left <- findOverlaps(junction_SJA.left, jncRanges, type="end")
        ol_A_from <- append(ol_A_from,
                            as.character(junction_SJA.left$id[ol_A.left@from]))
        if(exists("junctions_A")){
            junctions_A <- c(junctions_A, jncRanges[ol_A.left@to])
        }else{
            junctions_A <- jncRanges[ol_A.left@to]
        }
    }

    junctions_A$whippet_id <- ol_A_from
    junctions_A$search <- exonRanges$search[match(junctions_A$whippet_id, exonRanges$id)]
    junctions_A$set <- "A"

    if(type %in% c("AA","AD")){

        # junction B only required if AA/AD
        junction_SJB.right <- exonRanges[exonRanges$search=="right"]
        junction_SJB.left <- exonRanges[exonRanges$search=="left"]

        # same for B junctions
        ol_B_from <- vector()
        if(length(junction_SJB.right) > 0){
            start(junction_SJB.right) <- end(junction_SJB.right)
            ol_B.right <- findOverlaps(junction_SJB.right, jncRanges, type="start")
            ol_B_from <- append(ol_B_from,
                                as.character(junction_SJB.right$id[ol_B.right@from]))
            junctions_B <- jncRanges[ol_B.right@to]
        }
        if(length(junction_SJB.left) > 0){
            end(junction_SJB.left) <- end(junction_SJB.left) +1
            start(junction_SJB.left) <- end(junction_SJB.left)
            ol_B.left <- findOverlaps(junction_SJB.left, jncRanges, type="end")
            ol_B_from <- append(ol_B_from,
                                as.character(junction_SJB.left$id[ol_B.left@from]))
            if(exists("junctions_B")){
                junctions_B <- c(junctions_B, jncRanges[ol_B.left@to])
            }else{
                junctions_B <- jncRanges[ol_B.left@to]
            }
        }
        junctions_B$whippet_id <- ol_B_from
        junctions_B$search <- exonRanges$search[match(junctions_B$whippet_id, exonRanges$id)]
        junctions_B$set <- "B"
        junctions <- c(junctions_A, junctions_B)
    }

    if(type %in% c("AF","AL")){
        junctions_A.left <- junctions_A[junctions_A$search=="left"]
        junctions_A.right <- junctions_A[junctions_A$search=="right"]

        if(length(junctions_A.left) > 0){
            end(junctions_A.left) <- start(junctions_A.left)
            ol_C.left <- findOverlaps(junctions_A.left, jncRanges, type="start")
            junctions_C.left <- jncRanges[ol_C.left@to]
            junctions_C.left$whippet_id <- junctions_A.left$whippet_id[ol_C.left@from]
            junctions_C.left$search <- junctions_A.left$search[ol_C.left@from]
            ol <- findOverlaps(junctions_C.left, junctions_A, type="equal")
            if(length(ol) > 0){
                junctions_C.left <- junctions_C.left[-ol@from]
            }
            junctions_C <- junctions_C.left
        }
        if(length(junctions_A.right) > 0){
            start(junctions_A.right) <- end(junctions_A.right)
            ol_C.right <- findOverlaps(junctions_A.right, jncRanges, type="end")
            junctions_C.right <- jncRanges[ol_C.right@to]
            junctions_C.right$whippet_id <- junctions_A.right$whippet_id[ol_C.right@from]
            junctions_C.right$search <- junctions_A.right$search[ol_C.right@from]
            ol <- findOverlaps(junctions_C.right, junctions_A, type="equal")
            if(length(ol) > 0){
                junctions_C.right <- junctions_C.right[-ol@from]
            }
            if(exists("junctions_C")){
                junctions_C <- c(junctions_C, junctions_C.right)
            }else{
                junctions_C <- junctions_C.right
            }
        }

        junctions_C$set <- "C"
        junctions <- c(junctions_A, junctions_C)
    }

    keep <- which(width(junctions) > 2)

    # replace junction codes
    if(type %in% c("AA", "AD")){
        junctions$set[which((junctions$set=="A" & as.logical(strand(junctions) == "+")) |
                            (junctions$set=="B" & as.logical(strand(junctions) == "-"))  )] <- "X"
        junctions$set[which((junctions$set=="A" & as.logical(strand(junctions) == "-")) |
                             junctions$set=="B" & as.logical(strand(junctions) == "+") )] <- "Y"
    }
    if(type %in% c("AF", "AL")){
        junctions$set[which(junctions$set=="A")] <- "X"
        junctions$set[which(junctions$set=="C")] <- "Y"
    }

    junctions <- junctions[keep]
    return(junctions)
}

#' Find transcripts containing/overlapping junctions and replace them with alternative junctions
#'
#' @param junctionPairs GRanges object with alternative Whippet junctions. Generated by findJunctionPairs()
#' @param exonRanges GRanges object with Whippet event coordinates
#' @param type type of Whippet event (AA/AD/AF/AL). Note only one event type should be processed at a time.
#' @return GRanges object with transcripts containing alternative junctions.
#' @export
#' @import GenomicRanges
#' @examples
#' @author Beth Signal
replaceJunction <- function(junctionPairs, exonRanges, gtf.exons, type=NA){

    junctionPairs$type <- type
    range <- junctionPairs

    if(type %in% c("AA", "AD")){
        ## find exons that use/overlap the junction - at the side where it's alternative
        end(range)[which(range$seach=="right")] <- start(range)[which(range$seach=="right")]
        start(range)[which(range$seach=="left")] <- end(range)[which(range$seach=="left")]

        ol_junction <- findOverlaps(range, gtf.exons)
        ol_junction <- cbind(as.data.frame(ol_junction),
                             transcript_id=gtf.exons$transcript_id[ol_junction@to])

        ## table of transcripts overlapping the junction
        # tid: transcript id
        tid_table <- as.data.frame(table(ol_junction$queryHits, ol_junction$transcript_id))
        tid_table <- tid_table[tid_table$Freq > 0,]
        colnames(tid_table)[1:2] <- c("from_index","to_transcript_id")
        tids <- unique(tid_table$to_transcript_id)
        tid_table$junction_id <- range$id[tid_table$from_index]
        ## new transcript id -- unique if different junctions are going to be used in same transcript base
        tid_table$new_transcript_id <- paste0(tid_table$to_transcript_id,"+AS ",tid_table$junction_id)

        ## all transcripts for structural altercations
        gtf_transcripts <- gtf.exons[gtf.exons$transcript_id %in% tids]
        mcols(gtf_transcripts) <- mcols(gtf_transcripts)[,c('gene_id','transcript_id','exon_number')]
        m <- match(gtf_transcripts$transcript_id, tid_table$to_transcript_id)
        # add new transcript id
        gtf_transcripts$new_transcript_id <-
            paste0(gtf_transcripts$transcript_id,"+AS ",range$id[tid_table$from_index[m]])
        gtf_transcripts$new_transcript_id_exnum <- paste0(gtf_transcripts$new_transcript_id, "_", as.numeric(gtf_transcripts$exon_number))

        # duplicate core transcripts if needed
        needs_dup <- which(!(tid_table$new_transcript_id %in% gtf_transcripts$new_transcript_id))
        if(length(needs_dup) > 0){
            gtf_transcripts_add <- gtf_transcripts[gtf_transcripts$transcript_id %in% tid_table$to_transcript_id[needs_dup]]
        }
        while(length(needs_dup) > 0){
            gtf_transcripts_add <- gtf_transcripts_add[gtf_transcripts_add$transcript_id %in% tid_table$to_transcript_id[needs_dup]]
            m <- match(gtf_transcripts_add$transcript_id, tid_table$to_transcript_id[needs_dup])
            gtf_transcripts_add$new_transcript_id <- paste0(gtf_transcripts_add$transcript_id,"+AS ",tid_table$junction_id[needs_dup][m])
            gtf_transcripts <- c(gtf_transcripts, gtf_transcripts_add)
            needs_dup <- which(!(tid_table$new_transcript_id %in% gtf_transcripts$new_transcript_id))
        }

        gtf_transcripts$from <- unlist(lapply(str_split(gtf_transcripts$new_transcript_id, "AS "),"[[",2))
        gtf_transcripts <- gtf_transcripts[order(gtf_transcripts$transcript_id, start(gtf_transcripts))]

        ## alter exons hitting the junctions so they all break at the same place
        # range is at the alt. points defined in exonRanges
        range <- junctionPairs
        end(range) <- start(range)

        ol_left <- as.data.frame(findOverlaps(range, gtf_transcripts))
        ol_left$from_id <- range$id[ol_left$queryHits]
        ol_left$to_id <- gtf_transcripts$from[ol_left$subjectHits]
        ol_left <- ol_left[ol_left$from_id == ol_left$to_id,]

        # fix the end of the left transcript exons
        exons_left <- gtf_transcripts[ol_left$subjectHits]
        end(exons_left) <- end(range[ol_left$queryHits])

        # now the right side
        range <- junctionPairs
        start(range) <- end(range)

        ol_right <- as.data.frame(findOverlaps(range, gtf_transcripts))
        ol_right$from_id <- range$id[ol_right$queryHits]
        ol_right$to_id <- gtf_transcripts$from[ol_right$subjectHits]
        ol_right <- ol_right[ol_right$from_id == ol_right$to_id,]

        # fix the start of the right exons
        exons_right <- gtf_transcripts[ol_right$subjectHits]
        start(exons_right) <- end(range[ol_right$queryHits])


        m <- match(exons_left$new_transcript_id, exons_right$new_transcript_id)
        exons_left <- exons_left[which(!is.na(m))]
        exons_right <- exons_right[m[which(!is.na(m))]]

        exons_glued <- exons_left
        end(exons_glued) <- end(exons_right)

        # replacement exon pairs
        gtf_transcripts_replacement <- c(exons_left,exons_right)

        # remove replaced exons from gtf
        gtf_transcripts_altered <- gtf_transcripts[gtf_transcripts$new_transcript_id %in%
                                                       gtf_transcripts_replacement$new_transcript_id]


        ol <- as.data.frame(findOverlaps(exons_glued, gtf_transcripts_altered))
        ol$from_id <- exons_glued$new_transcript_id[ol$queryHits]
        ol$to_id <- gtf_transcripts_altered$new_transcript_id[ol$subjectHits]
        ol <- ol[ol$from_id == ol$to_id,]
        gtf_transcripts_altered <- gtf_transcripts_altered[-unique(ol$subjectHits)]

        # add together
        gtf_transcripts_altered <- c(gtf_transcripts_altered, gtf_transcripts_replacement)
        gtf_transcripts_altered <- gtf_transcripts_altered[order(gtf_transcripts_altered$new_transcript_id,
                                                                 start(gtf_transcripts_altered))]

        gtf_transcripts_altered$set <- range$set[match(gtf_transcripts_altered$from, range$id)]
        gtf_transcripts_altered$whippet_id <- junctionPairs$whippet_id[match(gtf_transcripts_altered$from, junctionPairs$id)]

    }else if(type %in% c("AF", "AL")){
        end(range)[which(range$search=="right")] <- start(range)[which(range$search=="right")]
        start(range)[which(range$search=="left")] <- end(range)[which(range$search=="left")]
        ol_firstLast.left <- findOverlaps(range[which(range$search=="left")], gtf.exons, type="start")
        ol_firstLast.right <- findOverlaps(range[which(range$search=="right")], gtf.exons, type="end")

        ol_firstLast.left <- cbind(as.data.frame(ol_firstLast.left),
                                   transcript_id=gtf.exons$transcript_id[ol_firstLast.left@to])
        ol_firstLast.right <- cbind(as.data.frame(ol_firstLast.right),
                                    transcript_id=gtf.exons$transcript_id[ol_firstLast.right@to])

        ol_firstLast <- rbind(ol_firstLast.left, ol_firstLast.right)

        exons_firstLast <- gtf.exons[ol_firstLast$subjectHits]
        exons_firstLast$set <- range$set[ol_firstLast$queryHits]
        exons_firstLast$search <- range$search[ol_firstLast$queryHits]

        exons_firstLast$junction_id <- range$id[ol_firstLast$queryHits]
        new_id.left <- paste0(seqnames(exons_firstLast),":",
                              start(junctionPairs)[ol_firstLast$queryHits],"-",
                              end(junctionPairs)[ol_firstLast$queryHits],"+",
                              end(exons_firstLast))
        new_id.right <- paste0(seqnames(exons_firstLast),":",
                               start(junctionPairs)[ol_firstLast$queryHits],"-",
                               end(junctionPairs)[ol_firstLast$queryHits],"+",
                               start(exons_firstLast))
        exons_firstLast$new_id[which(exons_firstLast$search=="left")] <- new_id.left[which(exons_firstLast$search=="left")]
        exons_firstLast$new_id[which(exons_firstLast$search=="right")] <- new_id.right[which(exons_firstLast$search=="right")]

        m <- match(exons_firstLast$junction_id, junctionPairs$id)
        junctionPairs <- junctionPairs[m]
        range <- junctionPairs
        range$id <- exons_firstLast$new_id
        end(range)[which(range$search=="left")] <- start(range)[which(range$search=="left")]
        start(range)[which(range$search=="right")] <- end(range)[which(range$search=="right")]

        ol_junction <- findOverlaps(range, gtf.exons)
        ol_junction <- cbind(as.data.frame(ol_junction),
                             transcript_id=gtf.exons$transcript_id[ol_junction@to])

        ## table of transcripts overlapping the junction
        # tid: transcript id
        tid_table <- as.data.frame(table(ol_junction$queryHits, ol_junction$transcript_id))
        tid_table <- tid_table[tid_table$Freq > 0,]
        colnames(tid_table)[1:2] <- c("from_index","to_transcript_id")
        tids <- unique(tid_table$to_transcript_id)

        tid_table$junction_id <- range$id[tid_table$from_index]
        ## new transcript id -- unique if different junctions are going to be used in same transcript base
        tid_table$new_transcript_id <- paste0(tid_table$to_transcript_id,"+AS ",tid_table$junction_id)
        ## all transcripts for structural altercations
        gtf_transcripts <- gtf.exons[gtf.exons$transcript_id %in% tids]
        mcols(gtf_transcripts) <- mcols(gtf_transcripts)[,c('gene_id','transcript_id','exon_number')]
        m <- match(gtf_transcripts$transcript_id, tid_table$to_transcript_id)
        # add new transcript id
        gtf_transcripts$new_transcript_id <-
            paste0(gtf_transcripts$transcript_id,"+AS ",range$id[tid_table$from_index[m]])
        gtf_transcripts$new_transcript_id_exnum <- paste0(gtf_transcripts$new_transcript_id, "_", as.numeric(gtf_transcripts$exon_number))

        # duplicate core transcripts if needed
        needs_dup <- which(!(tid_table$new_transcript_id %in% gtf_transcripts$new_transcript_id))
        if(length(needs_dup) > 0){
            gtf_transcripts_add <- gtf_transcripts[gtf_transcripts$transcript_id %in% tid_table$to_transcript_id[needs_dup]]
        }
        while(length(needs_dup) > 0){
            gtf_transcripts_add <- gtf_transcripts_add[gtf_transcripts_add$transcript_id %in% tid_table$to_transcript_id[needs_dup]]
            m <- match(gtf_transcripts_add$transcript_id, tid_table$to_transcript_id[needs_dup])
            gtf_transcripts_add$new_transcript_id <- paste0(gtf_transcripts_add$transcript_id,"+AS ",tid_table$junction_id[needs_dup][m])
            gtf_transcripts <- c(gtf_transcripts, gtf_transcripts_add)
            needs_dup <- which(!(tid_table$new_transcript_id %in% gtf_transcripts$new_transcript_id))
        }

        gtf_transcripts$from <- unlist(lapply(str_split(gtf_transcripts$new_transcript_id, "AS "),"[[",2))
        gtf_transcripts <- gtf_transcripts[order(gtf_transcripts$transcript_id, start(gtf_transcripts))]

        gtf_transcripts <- gtf_transcripts[gtf_transcripts$transcript_id %in% exons_firstLast$transcript_id]
        gtf_transcripts$new_transcript_id_exnum <- paste0(gtf_transcripts$new_transcript_id, "_", as.numeric(gtf_transcripts$exon_number))

        range <- junctionPairs
        range$id <- exons_firstLast$new_id

        ## find exons that use/overlap the junction - at the side where it's alternative
        end(range)[which(range$search=="left")] <- start(range)[which(range$search=="left")]
        start(range)[which(range$search=="right")] <- end(range)[which(range$search=="right")]

        ### Same used junction replacement
        w_left <- which(range$search=="left")
        ol_left <- as.data.frame(findOverlaps(range[w_left], gtf_transcripts))
        ol_left$from_id <- range$id[w_left][ol_left$queryHits]
        ol_left$to_id <- gtf_transcripts$from[ol_left$subjectHits]
        ol_left <- ol_left[ol_left$from_id == ol_left$to_id,]

        # fix the end of the left transcript exons
        exons_left <- gtf_transcripts[ol_left$subjectHits]
        end(exons_left) <- end(range[ol_left$queryHits])

        w_right <- which(range$search=="right")
        ol_right <- as.data.frame(findOverlaps(range[w_right], gtf_transcripts))
        ol_right$from_id <- range$id[w_right][ol_right$queryHits]
        ol_right$to_id <- gtf_transcripts$from[ol_right$subjectHits]
        ol_right <- ol_right[ol_right$from_id == ol_right$to_id,]

        # fix the end of the right transcript exons
        exons_right <- gtf_transcripts[ol_right$subjectHits]
        start(exons_right) <- start(range[ol_right$queryHits])

        junctionReplacementExons <- c(exons_left, exons_right)
        junctionReplacementExons$set <- range$set[match(junctionReplacementExons$from, range$id)]

        keep <- which(gtf_transcripts$new_transcript_id %in% replacementExonsFirstLast$new_transcript_id)
        gtf_transcripts_altered <- gtf_transcripts[keep]
        gtf_transcripts_altered$set <- range$set[match(gtf_transcripts_altered$from, range$id)]

        ### First/last exon replacement

        m <- match(junctionReplacementExons$from, exons_firstLast$new_id)
        replacementExonsFirstLast <- junctionReplacementExons
        ranges(replacementExonsFirstLast) <- ranges(exons_firstLast[m])


        # remove anything after first/last
        if(type=="AF"){
            back <- 0
            n <- which(gtf_transcripts_altered$new_transcript_id_exnum %in% junctionReplacementExons$new_transcript_id_exnum)
            ids <- gtf_transcripts_altered$new_transcript_id[n]
            exon_nums <- as.numeric(gtf_transcripts_altered$exon_number[n])
            new_transcript_id_exnum <- paste0(ids, "_", exon_nums-back)
            m <- match(new_transcript_id_exnum, gtf_transcripts_altered$new_transcript_id_exnum)
            m <- m[!is.na(m)]
            while(length(m) > 0){
                gtf_transcripts_altered <- gtf_transcripts_altered[-m]
                back <- back + 1
                new_transcript_id_exnum <- paste0(ids, "_", exon_nums-back)
                m <- match(new_transcript_id_exnum, gtf_transcripts_altered$new_transcript_id_exnum)
                m <- m[!is.na(m)]
            }
        }
        if(type=="AL"){
            fwd <- 0
            n <- which(gtf_transcripts_altered$new_transcript_id_exnum %in% junctionReplacementExons$new_transcript_id_exnum)
            ids <- gtf_transcripts_altered$new_transcript_id[n]
            exon_nums <- as.numeric(gtf_transcripts_altered$exon_number[n])
            new_transcript_id_exnum <- paste0(ids, "_", exon_nums+fwd)
            m <- match(new_transcript_id_exnum, gtf_transcripts_altered$new_transcript_id_exnum)
            m <- m[!is.na(m)]
            while(length(m) > 0){
                gtf_transcripts_altered <- gtf_transcripts_altered[-m]
                fwd <- fwd + 1
                new_transcript_id_exnum <- paste0(ids, "_", exon_nums+fwd)
                m <- match(new_transcript_id_exnum, gtf_transcripts_altered$new_transcript_id_exnum)
                m <- m[!is.na(m)]
            }
        }

        # remove overlapping exons
        longRange <- replacementExonsFirstLast
        range_df <- data.frame(start_1 = start(replacementExonsFirstLast),
                               start_2 = start(junctionReplacementExons),
                               end_1 = end(replacementExonsFirstLast),
                               end_2 = end(junctionReplacementExons))
        start(longRange) <- apply(range_df[,1:2], 1, min)
        end(longRange) <- apply(range_df[,3:4], 1, max)

        ol <- as.data.frame(findOverlaps(longRange, gtf_transcripts_altered))
        ol$from_id <- longRange$new_transcript_id[ol$queryHits]
        ol$to_id <- gtf_transcripts_altered$new_transcript_id[ol$subjectHits]
        ol <- ol[which(ol$from_id == ol$to_id),]
        if(dim(ol)[1] > 0){
            gtf_transcripts_altered <- (gtf_transcripts_altered[-unique(ol$subjectHits)])
        }
        gtf_transcripts_altered <- c(gtf_transcripts_altered,
                                     junctionReplacementExons,
                                     replacementExonsFirstLast)

        #redo exon numbering
        gtf_transcripts_altered <- gtf_transcripts_altered[order(gtf_transcripts_altered$new_transcript_id,
                                                                 start(gtf_transcripts_altered))]
        tab <- as.data.frame(table(gtf_transcripts_altered$new_transcript_id))
        tab$strand <- as.character(strand(gtf_transcripts_altered[match(tab$Var1,
                                                                        gtf_transcripts_altered$new_transcript_id)]))
        gtf_transcripts_altered$exon_number <- unlist(apply(tab, 1, function(x) if(x[3] == "-"){c(x[2]:1)}else{c(1:x[2])}))
        gtf_transcripts_altered <- gtf_transcripts_altered[order(gtf_transcripts_altered$new_transcript_id,
                                                                 start(gtf_transcripts_altered))]

        gtf_transcripts_altered <- gtf_transcripts_altered[order(gtf_transcripts_altered$new_transcript_id,
                                                                 start(gtf_transcripts_altered))]

        gtf_transcripts_altered$set <- range$set[match(gtf_transcripts_altered$from, range$id)]
        mcols(gtf_transcripts_altered) <- mcols(gtf_transcripts_altered)[,match(c('gene_id',"transcript_id","exon_number","from","set"),
                                                                                colnames(mcols(gtf_transcripts_altered)))]
        colnames(mcols(gtf_transcripts_altered))[4] <- "whippet_id"
    }


    #gtf_transcripts_altered$transcript_id <- gtf_transcripts_altered$new_transcript_id
    #mcols(gtf_transcripts_altered) <- mcols(gtf_transcripts_altered)[,c('gene_id','transcript_id','exon_number','whippet_id','set')]

    return(gtf_transcripts_altered)
}


