#' Format Whippet junctions as a GRanges object
#' @param junctions data.frame containing junction location information. May be generated by readWhippetJNCfiles()
#' @return GRanges object with junctions
#' @export
#' @import GenomicRanges
#' @examples
#' @author Beth Signal
#' @examples
#' whippetFiles <- list.files(system.file("extdata","whippet/", package = "GeneStructureTools"), full.names = TRUE)
#' jncFiles <- whippetFiles[grep(".jnc", whippetFiles)]
#' whippetJNC <- readWhippetJNCfiles(jncFiles)
#' junctionCoords <- formatJunctions(whippetJNC)
formatJunctions <- function(junctions){
    jncCoords <- GRanges(seqnames=Rle(junctions$chrom),
                          ranges=IRanges(start=as.numeric(junctions$start),
                                         end=as.numeric(junctions$end)),
                          strand=junctions$strand, id=junctions$id)
    return(jncCoords)
}


#' Find alternative junctions for Whippet alternative splicing events
#'
#' Find junctions that pair with each end of an AA (alt. acceptor) or AD (alt. donor) whippet range
#' Find junctions that pair with the upsteam/downstream exon of an AF (alt. first exon) or an AL (alt. last exon)
#' @param eventCoords GRanges object with Whippet event coordinates
#' @param jncCoords GRanges object with Whippet junctions. Generated by formatJunctions()
#' @param type type of Whippet event (AA/AD/AF/AL). Note only one event type should be processed at a time.
#' @return GRanges object with alternative junctions. Each event should have a set of X (for which the psi measurement is reported) junctions, and alternative Y junctions.
#' @export
#' @import GenomicRanges
#' @examples
#' @author Beth Signal
#' whippetFiles <- list.files(system.file("extdata","whippet/", package = "GeneStructureTools"), full.names = TRUE)
#' jncFiles <- whippetFiles[grep(".jnc", whippetFiles)]
#' whippetJNC <- readWhippetJNCfiles(jncFiles)
#' junctionCoords <- formatJunctions(whippetJNC)
#' diffFiles <- whippetFiles[grep(".diff", whippetFiles)]
#' whippetDiffSplice <- readWhippetDIFFfiles(diffFiles)
#' whippetCoords <- formatWhippetEvents(whippetDiffSplice)
#'
#' event.altAce <- whippetDiffSplice[which(whippetDiffSplice$type=="AA")[1],]
#' coords.altAce <- whippetCoords[whippetCoords$id %in% event.altAce$coord]
#' jncPairs.altAce <- findJunctionPairs(coords.altAce,junctionCoords, type="AA")
#'
#' event.altDon <- whippetDiffSplice[which(whippetDiffSplice$type=="AD")[1],]
#' coords.altDon <- whippetCoords[whippetCoords$id %in% event.altDon$coord]
#' jncPairs.altDon <- findJunctionPairs(coords.altDon,junctionCoords, type="AD")
#'
#' event.altFirst <- whippetDiffSplice[which(whippetDiffSplice$type=="AF")[1],]
#' coords.altFirst <- whippetCoords[whippetCoords$id %in% event.altFirst$coord]
#' jncPairs.altFirst <- findJunctionPairs(coords.altFirst,junctionCoords, type="AF")
#'
#' event.altLast <- whippetDiffSplice[which(whippetDiffSplice$type=="AL")[1],]
#' coords.altLast <- whippetCoords[whippetCoords$id %in% event.altLast$coord]
#' jncPairs.altLast <- findJunctionPairs(coords.altLast,junctionCoords, type="AL")
findJunctionPairs <- function(eventCoords, jncCoords, type=NA){

    eventCoords$type <- type
    # search for alternatives to the left or the right?
    eventCoords$search <- "right"
    eventCoords$search[(eventCoords$type=="AA" & as.logical(strand(eventCoords) == '+'))|
                          (eventCoords$type=="AD" & as.logical(strand(eventCoords) == '-'))|
                          (eventCoords$type=="AF" & as.logical(strand(eventCoords) == '-'))|
                          (eventCoords$type=="AL" & as.logical(strand(eventCoords) == '+'))] <- "left"

    junction_SJA.right <- eventCoords[eventCoords$search=="right"]
    junction_SJA.left <- eventCoords[eventCoords$search=="left"]

    # right if AA&-
    # left if AA&+
    ol_A_from <- vector()
    if(length(junction_SJA.right) > 0){
        if(type %in% c("AA", "AD")){
            start(junction_SJA.right) <- start(junction_SJA.right) -1
            end(junction_SJA.right) <- start(junction_SJA.right)
        }else{
            start(junction_SJA.right) <- end(junction_SJA.right)
        }
        ol_A.right <- findOverlaps(junction_SJA.right, jncCoords, type="start")
        ol_A_from <- append(ol_A_from,
                            as.character(junction_SJA.right$id[ol_A.right@from]))
        junctions_A <- jncCoords[ol_A.right@to]
    }
    if(length(junction_SJA.left) > 0){
        end(junction_SJA.left) <- start(junction_SJA.left)
        ol_A.left <- findOverlaps(junction_SJA.left, jncCoords, type="end")
        ol_A_from <- append(ol_A_from,
                            as.character(junction_SJA.left$id[ol_A.left@from]))
        if(exists("junctions_A")){
            junctions_A <- c(junctions_A, jncCoords[ol_A.left@to])
        }else{
            junctions_A <- jncCoords[ol_A.left@to]
        }
    }

    junctions_A$whippet_id <- ol_A_from
    junctions_A$search <- eventCoords$search[match(junctions_A$whippet_id, eventCoords$id)]
    junctions_A$set <- "A"

    if(type %in% c("AA","AD")){

        # junction B only required if AA/AD
        junction_SJB.right <- eventCoords[eventCoords$search=="right"]
        junction_SJB.left <- eventCoords[eventCoords$search=="left"]

        # same for B junctions
        ol_B_from <- vector()
        if(length(junction_SJB.right) > 0){
            start(junction_SJB.right) <- end(junction_SJB.right)
            ol_B.right <- findOverlaps(junction_SJB.right, jncCoords, type="start")
            ol_B_from <- append(ol_B_from,
                                as.character(junction_SJB.right$id[ol_B.right@from]))
            junctions_B <- jncCoords[ol_B.right@to]
        }
        if(length(junction_SJB.left) > 0){
            end(junction_SJB.left) <- end(junction_SJB.left) +1
            start(junction_SJB.left) <- end(junction_SJB.left)
            ol_B.left <- findOverlaps(junction_SJB.left, jncCoords, type="end")
            ol_B_from <- append(ol_B_from,
                                as.character(junction_SJB.left$id[ol_B.left@from]))
            if(exists("junctions_B")){
                junctions_B <- c(junctions_B, jncCoords[ol_B.left@to])
            }else{
                junctions_B <- jncCoords[ol_B.left@to]
            }
        }
        junctions_B$whippet_id <- ol_B_from
        junctions_B$search <- eventCoords$search[match(junctions_B$whippet_id, eventCoords$id)]
        junctions_B$set <- "B"
        junctions <- c(junctions_A, junctions_B)
    }

    if(type %in% c("AF","AL")){
        junctions_A.left <- junctions_A[junctions_A$search=="left"]
        junctions_A.right <- junctions_A[junctions_A$search=="right"]

        if(length(junctions_A.left) > 0){
            end(junctions_A.left) <- start(junctions_A.left)
            ol_C.left <- findOverlaps(junctions_A.left, jncCoords, type="start")
            junctions_C.left <- jncCoords[ol_C.left@to]
            junctions_C.left$whippet_id <- junctions_A.left$whippet_id[ol_C.left@from]
            junctions_C.left$search <- junctions_A.left$search[ol_C.left@from]
            ol <- findOverlaps(junctions_C.left, junctions_A, type="equal")
            if(length(ol) > 0){
                junctions_C.left <- junctions_C.left[-ol@from]
            }
            junctions_C <- junctions_C.left
        }
        if(length(junctions_A.right) > 0){
            start(junctions_A.right) <- end(junctions_A.right)
            ol_C.right <- findOverlaps(junctions_A.right, jncCoords, type="end")
            junctions_C.right <- jncCoords[ol_C.right@to]
            junctions_C.right$whippet_id <- junctions_A.right$whippet_id[ol_C.right@from]
            junctions_C.right$search <- junctions_A.right$search[ol_C.right@from]
            ol <- findOverlaps(junctions_C.right, junctions_A, type="equal")
            if(length(ol) > 0){
                junctions_C.right <- junctions_C.right[-ol@from]
            }
            if(exists("junctions_C")){
                junctions_C <- c(junctions_C, junctions_C.right)
            }else{
                junctions_C <- junctions_C.right
            }
        }

        junctions_C$set <- "C"
        junctions <- c(junctions_A, junctions_C)
    }

    keep <- which(width(junctions) > 2)

    # replace junction codes
    if(type %in% c("AA", "AD")){
        junctions$set[which((junctions$set=="A" & as.logical(strand(junctions) == "+")) |
                            (junctions$set=="B" & as.logical(strand(junctions) == "-"))  )] <- "X"
        junctions$set[which((junctions$set=="A" & as.logical(strand(junctions) == "-")) |
                             junctions$set=="B" & as.logical(strand(junctions) == "+") )] <- "Y"
    }
    if(type %in% c("AF", "AL")){
        junctions$set[which(junctions$set=="A")] <- "X"
        junctions$set[which(junctions$set=="C")] <- "Y"
    }

    junctions <- junctions[keep]
    return(junctions)
}

#' Find transcripts containing/overlapping junctions and replace them with alternative junctions
#'
#' @param junctionPairs GRanges object with alternative Whippet junctions. Generated by findJunctionPairs()
#' @param eventCoords GRanges object with Whippet event coordinates
#' @param type type of Whippet event (AA/AD/AF/AL). Note only one event type should be processed at a time.
#' @return GRanges object with transcripts containing alternative junctions.
#' @export
#' @import GenomicRanges
#' @examples
#' @author Beth Signal
#' @examples
#' whippetFiles <- list.files(system.file("extdata","whippet/", package = "GeneStructureTools"), full.names = TRUE)
#' jncFiles <- whippetFiles[grep(".jnc", whippetFiles)]
#' whippetJNC <- readWhippetJNCfiles(jncFiles)
#' junctionCoords <- formatJunctions(whippetJNC)
#' diffFiles <- whippetFiles[grep(".diff", whippetFiles)]
#' whippetDiffSplice <- readWhippetDIFFfiles(diffFiles)
#' whippetCoords <- formatWhippetEvents(whippetDiffSplice)
#' gtf <- rtracklayer::import(system.file("extdata","example_gtf.gtf",
#' package = "GeneStructureTools"))
#' gtf.exons <- gtf[gtf$type=="exon"]
#'
#' event.altAce <- whippetDiffSplice[which(whippetDiffSplice$type=="AA")[1],]
#' coords.altAce <- whippetCoords[whippetCoords$id %in% event.altAce$coord]
#' jncPairs.altAce <- findJunctionPairs(coords.altAce,junctionCoords, type="AA")
#' transcripts.altAce <- replaceJunction(jncPairs.altAce, coords.altAce, gtf.exons, type="AA")

#' event.altDon <- whippetDiffSplice[which(whippetDiffSplice$type=="AD")[1],]
#' coords.altDon <- whippetCoords[whippetCoords$id %in% event.altDon$coord]
#' jncPairs.altDon <- findJunctionPairs(coords.altDon,junctionCoords, type="AD")
#' transcripts.altDon <- replaceJunction(jncPairs.altDon, coords.altDon, gtf.exons, type="AD")
#'
#' event.altFirst <- whippetDiffSplice[which(whippetDiffSplice$type=="AF")[1],]
#' coords.altFirst <- whippetCoords[whippetCoords$id %in% event.altFirst$coord]
#' jncPairs.altFirst <- findJunctionPairs(coords.altFirst,junctionCoords, type="AF")
#' transcripts.altFirst <- replaceJunction(jncPairs.altFirst, coords.altFirst, gtf.exons, type="AF")
#'
#' event.altLast <- whippetDiffSplice[which(whippetDiffSplice$type=="AL")[1],]
#' coords.altLast <- whippetCoords[whippetCoords$id %in% event.altLast$coord]
#' jncPairs.altLast <- findJunctionPairs(coords.altLast,junctionCoords, type="AL")
#' transcripts.altLast <- replaceJunction(jncPairs.altLast, coords.altLast, gtf.exons, type="AL")
#'
replaceJunction <- function(junctionPairs, eventCoords, gtf.exons, type=NA){

    junctionPairs$type <- type
    range <- junctionPairs

    if(type %in% c("AA", "AD")){
        ## find exons that use/overlap the junction - at the side where it's alternative
        end(range)[which(range$search=="right")] <- start(range)[which(range$search=="right")]
        start(range)[which(range$search=="left")] <- end(range)[which(range$search=="left")]

        ol.junction <- findOverlaps(range, gtf.exons)
        ol.junction <- cbind(as.data.frame(ol.junction),
                             transcript_id=gtf.exons$transcript_id[ol.junction@to])

        ## table of transcripts overlapping the junction
        # tid: transcript id
        tidTable <- as.data.frame(table(ol.junction$queryHits, ol.junction$transcript_id))
        tidTable <- tidTable[tidTable$Freq > 0,]
        colnames(tidTable)[1:2] <- c("from_index","to_transcript_id")
        tids <- unique(tidTable$to_transcript_id)

        #all combinations of transcripts + junctions
        tidTable <- data.frame(from_index=rep(1:length(junctionPairs), each=length(tids)),
                                to_transcript_id=rep(tids, length(junctionPairs)),
                                Freq=1)

        tidTable$junction_id <- range$id[tidTable$from_index]
        ## new transcript id -- unique if different junctions are going to be used in same transcript base
        tidTable$new_transcript_id <- paste0(tidTable$to_transcript_id,"+AS"," ",tidTable$junction_id)

        ## all transcripts for structural altercations
        gtfTranscripts <- gtf.exons[gtf.exons$transcript_id %in% tids]
        mcols(gtfTranscripts) <- mcols(gtfTranscripts)[,c('gene_id','transcript_id','transcript_type','exon_id','exon_number')]
        m <- match(gtfTranscripts$transcript_id, tidTable$to_transcript_id)
        # add new transcript id
        gtfTranscripts$new_transcript_id <-
            paste0(gtfTranscripts$transcript_id,"+AS ",range$id[tidTable$from_index[m]])
        gtfTranscripts$new_transcript_id_exnum <- paste0(gtfTranscripts$new_transcript_id, "_", as.numeric(gtfTranscripts$exon_number))

        # duplicate core transcripts if needed
        needsDuplicated <- which(!(tidTable$new_transcript_id %in% gtfTranscripts$new_transcript_id))
        if(length(needsDuplicated) > 0){
            gtfTranscripts.add <- gtfTranscripts[gtfTranscripts$transcript_id %in% tidTable$to_transcript_id[needsDuplicated]]
        }
        while(length(needsDuplicated) > 0){
            gtfTranscripts.add <- gtfTranscripts.add[gtfTranscripts.add$transcript_id %in% tidTable$to_transcript_id[needsDuplicated]]
            m <- match(gtfTranscripts.add$transcript_id, tidTable$to_transcript_id[needsDuplicated])
            gtfTranscripts.add$new_transcript_id <- paste0(gtfTranscripts.add$transcript_id,"+AS ",tidTable$junction_id[needsDuplicated][m])
            gtfTranscripts <- c(gtfTranscripts, gtfTranscripts.add)
            needsDuplicated <- which(!(tidTable$new_transcript_id %in% gtfTranscripts$new_transcript_id))
        }

        gtfTranscripts$from <- unlist(lapply(str_split(gtfTranscripts$new_transcript_id, "AS "),"[[",2))
        gtfTranscripts <- gtfTranscripts[order(gtfTranscripts$transcript_id, start(gtfTranscripts))]

        ## alter exons hitting the junctions so they all break at the same place
        # range is at the alt. points defined in eventCoords
        range <- junctionPairs
        start(range) <- min(start(junctionPairs))
        end(range) <- max(start(junctionPairs))

        ol.left <- as.data.frame(findOverlaps(range, gtfTranscripts))
        ol.left$from_id <- range$id[ol.left$queryHits]
        ol.left$to_id <- gtfTranscripts$from[ol.left$subjectHits]
        ol.left <- ol.left[ol.left$from_id == ol.left$to_id,]

        # fix the end of the left transcript exons
        exons.left <- gtfTranscripts[ol.left$subjectHits]

        keep <- which(start(exons.left) < start(junctionPairs[ol.left$queryHits]))
        end(exons.left)[keep] <- start(junctionPairs[ol.left$queryHits])[keep]
        exons.left <- exons.left[keep]

        # now the right side
        range <- junctionPairs

        end(range) <- max(end(junctionPairs))
        start(range) <- min(end(junctionPairs))

        ol.right <- as.data.frame(findOverlaps(range, gtfTranscripts))
        ol.right$from_id <- range$id[ol.right$queryHits]
        ol.right$to_id <- gtfTranscripts$from[ol.right$subjectHits]
        ol.right <- ol.right[ol.right$from_id == ol.right$to_id,]

        # fix the start of the right exons
        exons.right <- gtfTranscripts[ol.right$subjectHits]

        keep <- which(end(exons.right) > end(junctionPairs[ol.right$queryHits]))
        start(exons.right)[keep] <- end(junctionPairs[ol.right$queryHits])[keep]
        exons.right <- exons.right[keep]

        m <- match(exons.left$new_transcript_id, exons.right$new_transcript_id)
        exons.left <- exons.left[which(!is.na(m))]
        exons.right <- exons.right[m[which(!is.na(m))]]

        exons.glued <- exons.left
        end(exons.glued) <- end(exons.right)

        # replacement exon pairs
        gtfTranscripts.replacement <- c(exons.left,exons.right)

        # remove replaced exons from gtf
        gtfTranscripts.altered <- gtfTranscripts[gtfTranscripts$new_transcript_id %in%
                                                       gtfTranscripts.replacement$new_transcript_id]


        ol <- as.data.frame(findOverlaps(exons.glued, gtfTranscripts.altered))
        ol$from_id <- exons.glued$new_transcript_id[ol$queryHits]
        ol$to_id <- gtfTranscripts.altered$new_transcript_id[ol$subjectHits]
        ol <- ol[ol$from_id == ol$to_id,]
        gtfTranscripts.altered <- gtfTranscripts.altered[-unique(ol$subjectHits)]

        # add together
        gtfTranscripts.altered <- c(gtfTranscripts.altered, gtfTranscripts.replacement)
        gtfTranscripts.altered <- gtfTranscripts.altered[order(gtfTranscripts.altered$new_transcript_id,
                                                                 start(gtfTranscripts.altered))]

        gtfTranscripts.altered$set <- range$set[match(gtfTranscripts.altered$from, range$id)]
        gtfTranscripts.altered$whippet_id <- junctionPairs$whippet_id[match(gtfTranscripts.altered$from, junctionPairs$id)]
        gtfTranscripts.altered$transcript_id <- paste0(gtfTranscripts.altered$transcript_id,
                                                            "+AS",type,gtfTranscripts.altered$set," ",
                                                            gtfTranscripts.altered$whippet_id)
        gtfTranscripts.altered$set <- paste0(type, "_",gtfTranscripts.altered$set)


    }else if(type %in% c("AF", "AL")){
        end(range)[which(range$search=="right")] <- start(range)[which(range$search=="right")]
        start(range)[which(range$search=="left")] <- end(range)[which(range$search=="left")]

        ol_firstLast.left <- findOverlaps(range, gtf.exons, type="start")
        ol_firstLast.right <- findOverlaps(range, gtf.exons, type="end")
        ol_firstLast.left <- cbind(as.data.frame(ol_firstLast.left),
                                   transcript_id=gtf.exons$transcript_id[ol_firstLast.left@to])
        ol_firstLast.left <- ol_firstLast.left[which(range$search[ol_firstLast.left$queryHits] == "left"),]

        ol_firstLast.right <- cbind(as.data.frame(ol_firstLast.right),
                                    transcript_id=gtf.exons$transcript_id[ol_firstLast.right@to])
        ol_firstLast.right <- ol_firstLast.right[which(range$search[ol_firstLast.right$queryHits] == "right"),]

        ol_firstLast <- rbind(ol_firstLast.left, ol_firstLast.right)

        ol_firstLast$search <- range$search[ol_firstLast$queryHits]


        exons_firstLast <- gtf.exons[ol_firstLast$subjectHits]
        exons_firstLast$set <- range$set[ol_firstLast$queryHits]
        exons_firstLast$search <- range$search[ol_firstLast$queryHits]

        exons_firstLast$junction_id <- range$id[ol_firstLast$queryHits]
        new_id.left <- paste0(seqnames(exons_firstLast),":",
                              start(junctionPairs)[ol_firstLast$queryHits],"-",
                              end(junctionPairs)[ol_firstLast$queryHits],"+",
                              end(exons_firstLast))
        new_id.right <- paste0(seqnames(exons_firstLast),":",
                               start(junctionPairs)[ol_firstLast$queryHits],"-",
                               end(junctionPairs)[ol_firstLast$queryHits],"+",
                               start(exons_firstLast))
        exons_firstLast$new_id <- NA
        exons_firstLast$new_id[which(exons_firstLast$search=="left")] <- new_id.left[which(exons_firstLast$search=="left")]
        exons_firstLast$new_id[which(exons_firstLast$search=="right")] <- new_id.right[which(exons_firstLast$search=="right")]

        m <- match(exons_firstLast$junction_id, junctionPairs$id)
        junctionPairs <- junctionPairs[m]
        range <- junctionPairs
        range$id <- exons_firstLast$new_id
        end(range)[which(range$search=="left")] <- start(range)[which(range$search=="left")]
        start(range)[which(range$search=="right")] <- end(range)[which(range$search=="right")]

        ol.junction <- findOverlaps(range, gtf.exons)
        ol.junction <- cbind(as.data.frame(ol.junction),
                             transcript_id=gtf.exons$transcript_id[ol.junction@to])

        ## table of transcripts overlapping the junction
        # tid: transcript id
        tidTable <- as.data.frame(table(ol.junction$queryHits, ol.junction$transcript_id))
        tidTable <- tidTable[tidTable$Freq > 0,]
        colnames(tidTable)[1:2] <- c("from_index","to_transcript_id")
        tids <- unique(tidTable$to_transcript_id)

        tidTable$junction_id <- range$id[tidTable$from_index]
        ## new transcript id -- unique if different junctions are going to be used in same transcript base
        tidTable$new_transcript_id <- paste0(tidTable$to_transcript_id,"+AS ",tidTable$junction_id)
        ## all transcripts for structural altercations
        gtfTranscripts <- gtf.exons[gtf.exons$transcript_id %in% tids]
        mcols(gtfTranscripts) <- mcols(gtfTranscripts)[,c('gene_id','transcript_id','transcript_type','exon_id','exon_number')]
        m <- match(gtfTranscripts$transcript_id, tidTable$to_transcript_id)
        # add new transcript id
        gtfTranscripts$new_transcript_id <-
            paste0(gtfTranscripts$transcript_id,"+AS ",range$id[tidTable$from_index[m]])
        gtfTranscripts$new_transcript_id_exnum <- paste0(gtfTranscripts$new_transcript_id, "_", as.numeric(gtfTranscripts$exon_number))

        # duplicate core transcripts if needed
        needsDuplicated <- which(!(tidTable$new_transcript_id %in% gtfTranscripts$new_transcript_id))
        if(length(needsDuplicated) > 0){
            gtfTranscripts.add <- gtfTranscripts[gtfTranscripts$transcript_id %in% tidTable$to_transcript_id[needsDuplicated]]
        }
        while(length(needsDuplicated) > 0){
            gtfTranscripts.add <- gtfTranscripts.add[gtfTranscripts.add$transcript_id %in% tidTable$to_transcript_id[needsDuplicated]]
            m <- match(gtfTranscripts.add$transcript_id, tidTable$to_transcript_id[needsDuplicated])
            gtfTranscripts.add$new_transcript_id <- paste0(gtfTranscripts.add$transcript_id,"+AS ",tidTable$junction_id[needsDuplicated][m])
            gtfTranscripts <- c(gtfTranscripts, gtfTranscripts.add)
            needsDuplicated <- which(!(tidTable$new_transcript_id %in% gtfTranscripts$new_transcript_id))
        }

        gtfTranscripts$from <- unlist(lapply(str_split(gtfTranscripts$new_transcript_id, "AS "),"[[",2))
        gtfTranscripts <- gtfTranscripts[order(gtfTranscripts$transcript_id, start(gtfTranscripts))]

        gtfTranscripts <- gtfTranscripts[gtfTranscripts$transcript_id %in% exons_firstLast$transcript_id]
        gtfTranscripts$new_transcript_id_exnum <- paste0(gtfTranscripts$new_transcript_id, "_", as.numeric(gtfTranscripts$exon_number))

        range <- junctionPairs
        range$id <- exons_firstLast$new_id

        ## find exons that use/overlap the junction - at the side where it's alternative
        end(range)[which(range$search=="left")] <- start(range)[which(range$search=="left")]
        start(range)[which(range$search=="right")] <- end(range)[which(range$search=="right")]

        ### Same used junction replacement
        ol.left <- as.data.frame(findOverlaps(range, gtfTranscripts))
        ol.left$from_id <- range$id[ol.left$queryHits]
        ol.left$to_id <- gtfTranscripts$from[ol.left$subjectHits]
        ol.left <- ol.left[ol.left$from_id == ol.left$to_id,]
        ol.left <- ol.left[which(range$search[ol.left$queryHits] == "left"),]

        # fix the end of the left transcript exons
        exons.left <- gtfTranscripts[ol.left$subjectHits]
        end(exons.left) <- end(range[ol.left$queryHits])

        ol.right <- as.data.frame(findOverlaps(range, gtfTranscripts))
        ol.right$from_id <- range$id[ol.right$queryHits]
        ol.right$to_id <- gtfTranscripts$from[ol.right$subjectHits]
        ol.right <- ol.right[ol.right$from_id == ol.right$to_id,]
        ol.right <- ol.right[which(range$search[ol.right$queryHits] == "right"),]

        # fix the end of the right transcript exons
        exons.right <- gtfTranscripts[ol.right$subjectHits]
        start(exons.right) <- start(range[ol.right$queryHits])

        junctionReplacementExons <- c(exons.left, exons.right)
        junctionReplacementExons$set <- range$set[match(junctionReplacementExons$from, range$id)]

        keep <- which(gtfTranscripts$new_transcript_id %in% junctionReplacementExons$new_transcript_id)
        gtfTranscripts.altered <- gtfTranscripts[keep]
        gtfTranscripts.altered$set <- range$set[match(gtfTranscripts.altered$from, range$id)]

        ### First/last exon replacement

        m <- match(junctionReplacementExons$from, exons_firstLast$new_id)
        replacementExonsFirstLast <- junctionReplacementExons
        ranges(replacementExonsFirstLast) <- ranges(exons_firstLast[m])


        # remove anything after first/last
        if(type=="AF"){
            back <- 0
            n <- which(gtfTranscripts.altered$new_transcript_id_exnum %in% junctionReplacementExons$new_transcript_id_exnum)
            ids <- gtfTranscripts.altered$new_transcript_id[n]
            exonNumbers <- as.numeric(gtfTranscripts.altered$exon_number[n])
            altTidExNum <- paste0(ids, "_", exonNumbers-back)
            m <- match(altTidExNum, gtfTranscripts.altered$new_transcript_id_exnum)
            m <- m[!is.na(m)]
            while(length(m) > 0){
                gtfTranscripts.altered <- gtfTranscripts.altered[-m]
                back <- back + 1
                altTidExNum <- paste0(ids, "_", exonNumbers-back)
                m <- match(altTidExNum, gtfTranscripts.altered$new_transcript_id_exnum)
                m <- m[!is.na(m)]
            }
        }
        if(type=="AL"){
            fwd <- 0
            n <- which(gtfTranscripts.altered$new_transcript_id_exnum %in% junctionReplacementExons$new_transcript_id_exnum)
            ids <- gtfTranscripts.altered$new_transcript_id[n]
            exonNumbers <- as.numeric(gtfTranscripts.altered$exon_number[n])
            altTidExNum <- paste0(ids, "_", exonNumbers+fwd)
            m <- match(altTidExNum, gtfTranscripts.altered$new_transcript_id_exnum)
            m <- m[!is.na(m)]
            while(length(m) > 0){
                gtfTranscripts.altered <- gtfTranscripts.altered[-m]
                fwd <- fwd + 1
                altTidExNum <- paste0(ids, "_", exonNumbers+fwd)
                m <- match(altTidExNum, gtfTranscripts.altered$new_transcript_id_exnum)
                m <- m[!is.na(m)]
            }
        }

        # remove overlapping exons
        longRange <- replacementExonsFirstLast
        rangeDF <- data.frame(start_1 = start(replacementExonsFirstLast),
                               start_2 = start(junctionReplacementExons),
                               end_1 = end(replacementExonsFirstLast),
                               end_2 = end(junctionReplacementExons))
        start(longRange) <- apply(rangeDF[,1:2], 1, min)
        end(longRange) <- apply(rangeDF[,3:4], 1, max)

        ol <- as.data.frame(findOverlaps(longRange, gtfTranscripts.altered))
        ol$from_id <- longRange$new_transcript_id[ol$queryHits]
        ol$to_id <- gtfTranscripts.altered$new_transcript_id[ol$subjectHits]
        ol <- ol[which(ol$from_id == ol$to_id),]
        if(dim(ol)[1] > 0){
            gtfTranscripts.altered <- (gtfTranscripts.altered[-unique(ol$subjectHits)])
        }
        gtfTranscripts.altered <- c(gtfTranscripts.altered,
                                     junctionReplacementExons,
                                     replacementExonsFirstLast)

        #redo exon numbering
        gtfTranscripts.altered <- gtfTranscripts.altered[order(gtfTranscripts.altered$new_transcript_id,
                                                                 start(gtfTranscripts.altered))]
        tab <- as.data.frame(table(gtfTranscripts.altered$new_transcript_id))
        tab$strand <- as.character(strand(gtfTranscripts.altered[match(tab$Var1,
                                                                        gtfTranscripts.altered$new_transcript_id)]))
        gtfTranscripts.altered$exon_number <- unlist(apply(tab, 1, function(x) if(x[3] == "-"){c(x[2]:1)}else{c(1:x[2])}))

        gtfTranscripts.altered <- gtfTranscripts.altered[order(gtfTranscripts.altered$new_transcript_id,
                                                                 start(gtfTranscripts.altered))]

        gtfTranscripts.altered$set <- range$set[match(gtfTranscripts.altered$from, range$id)]
        mcols(gtfTranscripts.altered) <- mcols(gtfTranscripts.altered)[,match(c('gene_id','transcript_id','transcript_type','exon_id','exon_number','from','set'),
                                                                                colnames(mcols(gtfTranscripts.altered)))]
        colnames(mcols(gtfTranscripts.altered))[6] <- "new_event_id"
        gtfTranscripts.altered$whippet_id <- range$whippet_id[match(gtfTranscripts.altered$new_event_id, range$id)]
        gtfTranscripts.altered$transcript_id <- paste0(gtfTranscripts.altered$transcript_id,
                                                            "+AS",type,gtfTranscripts.altered$set," ",
                                                            gtfTranscripts.altered$whippet_id)
        gtfTranscripts.altered$set <- paste0(type, "_",gtfTranscripts.altered$set)

    }

    #gtfTranscripts.altered$transcript_id <- gtfTranscripts.altered$new_transcript_id
    mcols(gtfTranscripts.altered) <- mcols(gtfTranscripts.altered)[,c('gene_id','transcript_id','transcript_type','exon_id','exon_number',
                                                                        'set','whippet_id')]
    gtfTranscripts.altered <- removeDuplicateTranscripts(gtfTranscripts.altered)

    return(gtfTranscripts.altered)
}
#' Remove transcript duplicates
#'
#' Removes Structural duplicates of transcripts in a GRanges object
#' Note that duplicates must have different transcript ids.
#' @param transcripts GRanges object with transcript structures in exon form
#' @return GRanges object with unique transcript structures in exon form
#' @export
#' @import GenomicRanges
#' @author Beth Signal
#' @examples
#' gtf <- rtracklayer::import(system.file("extdata","example_gtf.gtf",
#' package = "GeneStructureTools"))
#' gtf.exons <- gtf[gtf$type=="exon"]
#' gtf.exons.altName <- gtf.exons
#' gtf.exons.altName$transcript_id <- paste(gtf.exons.altName$transcript_id, "duplicated", sep="_")
#' gtf.exons.duplicated <- c(gtf.exons, gtf.exons.altName)
#' length(gtf.exons.duplicated)
#' gtf.exons.deduplicated <- removeDuplicateTranscripts(gtf.exons.duplicated)
#' length(gtf.exons.deduplicated)

removeDuplicateTranscripts <- function(transcripts){
    transcriptDF <- as.data.frame(transcripts)

    transcriptDF$startend <- with(transcriptDF, paste0(start,"-",end))
    transcriptRangePaste <- aggregate(startend ~ transcript_id, transcriptDF, function(x) paste0(x, collapse="+"))

    keep <- transcriptRangePaste$transcript_id[which(!duplicated(transcriptRangePaste$startend))]
    transcriptsFiltered <- transcripts[transcripts$transcript_id %in% keep]
    return(transcriptsFiltered)
}
