#' Evaluate changes to ORFs caused by alternative splicing
#' @param orfs_normal orf information for 'normal' transcripts. Generated by getOrfs()
#' @param orfs_alternative orf information for 'alternative' transcripts. Generated by getOrfs()
#' @param transcript_matches data.frame with all relationships to be tested
#' @return data.frame with orf changes
#' @export
#' @import stringr
#' @examples
#' @author Beth Signal
orfDiff <- function(orfs_normal, orfs_alternative, transcript_matches){
    orf_changes <- attrChangeAltSpliced(orfs_normal,orfs_skipped, attribute = "orf_length", useMax=TRUE)

    id <- paste0(transcript_matches$gene_id[match(orf_changes$spliced_id, transcript_matches$from)],"_", orf_changes$orf_length_normal_gene)
    m1 <- match(id, paste0(orfs_normal$gene_id, "_", orfs_normal$orf_length))

    id <- paste0(orf_changes$spliced_id, "_", orf_changes$orf_length)
    m2 <- match(id, paste0(unlist(lapply(str_split(orfs_skipped$id," "),"[[", 2)), "_", orfs_skipped$orf_length))

    x <- as.numeric(mapply(function(x,y) orfSimilarity(x,y),
                           orfs_normal$orf_sequence[m1],
                           orfs_skipped$orf_sequence[m2]))

    orf_changes$percent_orf_kept <- x

    orf_changes$max_percent_orf_kept <- orf_changes$orf_length / orf_changes$orf_length_normal_gene
    orf_changes$max_percent_orf_kept[orf_changes$max_percent_orf_kept > 1] <- 1

    orf_length_kept <- orf_changes$percent_orf_kept * orf_changes$orf_length_normal_gene
    orf_changes$orf_percent_new <- 1 - orf_length_kept / orf_changes$orf_length

    return(orf_changes)
}

#' Evaluate the change in an attribute between a set of 'normal' transcripts and 'alternative' transcripts
#' @param orfs_normal orf information for 'normal' transcripts. Generated by getOrfs()
#' @param orfs_alternative orf information for 'alternative' transcripts. Generated by getOrfs()
#' @param attribute attribute to compare
#' @param useMax use max as the summary function when multiple isoforms are aggregated? If FALSE, will use min instead.
#' @return data.frame with attribute changes
#' @export
#' @import stringr
#' @examples
#' @author Beth Signal
attrChangeAltSpliced <- function(orfs_normal, orfs_alternative, attribute="orf_length", useMax=TRUE){

    if(useMax){
        agg_fun <- max
    }else{
        agg_fun <- min
    }

    m <- match(c('gene_id', attribute), colnames(orfs_normal))
    orfs_normal_part <- orfs_normal[,m]

    attr_normal <- aggregate(. ~ gene_id, orfs_normal_part, agg_fun)

    # minimum/maximum attribute for all alternative transcripts (all frames)

    m <- match(c('id', attribute), colnames(orfs_alternative))
    orfs_alternative_part <- orfs_alternative[,m]
    attr_retained <- aggregate(. ~ id, orfs_alternative_part, agg_fun)

    colnames(attr_retained)[2] <- "attr"

    attr_retained$transcript_id <- unlist(lapply(str_split(attr_retained$id, " "), "[[", 1))
    attr_retained$transcript_id <- gsub("[-]","", gsub("[+]", "", gsub("EXON","", gsub("INTRON","", attr_retained$transcript_id))))
    attr_retained$spliced_id <- unlist(lapply(str_split(attr_retained$id, " "), "[[", 2))
    # match retained transcripts to normal transcripts

    m <- match(attr_retained$transcript_id, orfs_normal$id)
    attr_retained$normal_attr <- orfs_normal[m,attribute]

    m <- match(attr_retained$transcript_id, orfs_normal$id)
    m2 <- match(orfs_normal$gene_id[m], attr_normal$gene_id)
    attr_retained$normal_attr_gene <- attr_normal[m2,attribute]

    attr_retained_bySpliced <- aggregate(attr ~ spliced_id, attr_retained, agg_fun)
    attr_retained_bySpliced_normal <- aggregate(normal_attr_gene ~ spliced_id, attr_retained, agg_fun)

    m <- match(attr_retained_bySpliced$spliced_id, attr_retained_bySpliced_normal$spliced_id)

    attr_retained_bySpliced <- cbind(attr_retained_bySpliced,
                                     attr_normal_gene=attr_retained_bySpliced_normal[m,2])

    colnames(attr_retained_bySpliced) <- gsub("attr", attribute, colnames(attr_retained_bySpliced))

    return(attr_retained_bySpliced)
}

#' calculate percentage of orf_b contained in orf_a
#' @param orf_a character string of ORF amino acid sequence
#' @param orf_b character string of ORF amino acid sequence
#' @return percentage of orf_b contained in orf_a
#' @export
#' @import stringdist
#' @examples
#' @author Beth Signal
orfSimilarity <- function(orf_a, orf_b){
    ((nchar(orf_a) + nchar(orf_b) - stringdist::stringdist(orf_a ,orf_b, method = "lcs"))/2) / nchar(orf_a)
}
