#' Evaluate changes to ORFs caused by alternative splicing
#' @param orfs_x orf information for 'normal' transcripts. Generated by getOrfs()
#' @param orfs_y orf information for 'alternative' transcripts. Generated by getOrfs()
#' @return data.frame with orf changes
#' @export
#' @import stringr
#' @examples
#' @author Beth Signal
orfDiff <- function(orfs_x, orfs_y){
    orf_changes <- attrChangeAltSpliced(orfs_x, orfs_y, attribute = "orf_length",
                                        compareBy="gene",useMax = TRUE)


    orfs_y$transcript_id <- unlist(lapply(str_split(orfs_y$id, "[+]"),"[[", 1))
    orfs_y$spliced_id <- unlist(lapply(str_split(orfs_y$id, "AS "),"[[", 2))
    orf_changes$transcript_id <- orfs_y$transcript_id[match(orf_changes$id, orfs_y$spliced_id)]
    orf_changes$gene_id <- orfs_y$gene_id[match(orf_changes$id, orfs_y$spliced_id)]

    orfs_y$id_with_len <- paste0(orfs_y$spliced_id, "_", orfs_y$orf_length)
    orf_changes$id_orf_length_y <- paste0(orf_changes$id, "_",orf_changes$orf_length_bygroup_y)

    my <- match(orf_changes$id_orf_length_y, orfs_y$id_with_len)

    if(all(!grepl("AS ", orfs_x$id))){
        orfs_x$id_with_len <- paste0(orfs_x$gene_id, "_", orfs_x$orf_length)
        orf_changes$id_orf_length_x <- paste0(orf_changes$gene_id, "_",orf_changes$orf_length_bygroup_x)
    }else{
        orfs_x$spliced_id <- unlist(lapply(str_split(orfs_x$id, "AS "),"[[", 2))
        orfs_x$id_with_len <- paste0(orfs_x$spliced_id, "_", orfs_x$orf_length)
        orf_changes$id_orf_length_x <- paste0(orf_changes$id, "_",orf_changes$orf_length_bygroup_x)
    }

    mx <- match(orf_changes$id_orf_length_x, orfs_x$id_with_len)

    x <- as.numeric(mapply(function(x,y) orfSimilarity(x,y),
                           orfs_x$orf_sequence[mx],
                           orfs_y$orf_sequence[my]))

    orf_changes$percent_orf_kept <- x

    orf_changes$max_percent_orf_kept <- orf_changes$orf_length_bygroup_y /
        orf_changes$orf_length_bygroup_x
    orf_changes$max_percent_orf_kept[orf_changes$max_percent_orf_kept > 1] <- 1

    orf_length_kept <- orf_changes$percent_orf_kept * orf_changes$orf_length_bygroup_x
    orf_changes$orf_percent_new <- 1 - orf_length_kept / orf_changes$orf_length_bygroup_y
    orf_changes$transcript_id <- NULL
    orf_changes$gene_id <- NULL
    orf_changes$id_orf_length_x <- NULL
    orf_changes$id_orf_length_y <- NULL

    return(orf_changes)
}

#' Evaluate the change in an attribute between a set of 'normal' transcripts and 'alternative' transcripts
#' @param orfs_x orf information for 'normal' transcripts. Generated by getOrfs()
#' @param orfs_y orf information for 'alternative' transcripts. Generated by getOrfs()
#' @param attribute attribute to compare
#' @param compareBy compare by 'transcript' isoforms or by 'gene' groups
#' @param useMax use max as the summary function when multiple isoforms are aggregated? If FALSE, will use min instead.
#' @return data.frame with attribute changes
#' @export
#' @import stringr
#' @examples
#' @author Beth Signal
attrChangeAltSpliced <- function(orfs_x, orfs_y, attribute="orf_length",
                                 compareBy="transcript", useMax=TRUE){

    if(useMax){
        agg_fun <- max
    }else{
        agg_fun <- min
    }

    m <- match(c('id','gene_id', attribute), colnames(orfs_x))
    orfs_x_part <- orfs_x[,m]
    attr_x <- aggregate(. ~ id+gene_id, orfs_x_part, agg_fun)

    m <- match(c('id','gene_id', attribute), colnames(orfs_y))
    orfs_y_part <- orfs_y[,m]
    attr_y <- aggregate(. ~ id+gene_id, orfs_y_part, agg_fun)

    attr_y$as_group <- unlist(lapply(str_split(attr_y$id, " "), '[[', 2))

    if(all(!grepl("AS ", attr_x$id))){
        transcript_id <- unlist(lapply(str_split(attr_y$id, "[+]"), '[[', 1))
        m <- match(transcript_id, attr_x$id)
        attr_x <- attr_x[m,]
        attr_x$as_group <- attr_y$as_group
        attr_x$id <- attr_y$id
    }else{
        attr_x$as_group <- unlist(lapply(str_split(attr_x$id, " "), '[[', 2))
    }
    colnames(attr_x)[3] <- "attr"
    colnames(attr_y)[3] <- "attr"

    if(compareBy == "transcript"){
        m <- match(attr_x$id, attr_y$id)
        attr_comps <- data.frame(id=attr_x$id,
                                 attr_bygroup_x=attr_x$attr,
                                 attr_bygroup_y=attr_y$attr[m])

        m <- match(attr_y$id, attr_x$id)
        attr_comps_y <- data.frame(id=attr_y$id,
                                   attr_bygroup_x=attr_x$attr[m],
                                   attr_bygroup_y=attr_y$attr)

        add <- which(!(attr_comps_y$id %in% attr_comps$id))

        if(length(add) > 0){
            attr_comps <- rbind(attr_comps, attr_comps_y[add,])
        }
    }else if(compareBy=="gene"){
        attr_x2 <- aggregate(attr ~ as_group, attr_x, agg_fun)
        attr_y2 <- aggregate(attr ~ as_group, attr_y, agg_fun)

        m <- match(attr_x2$as_group, attr_y2$as_group)
        attr_comps <- data.frame(id=attr_x2[,1],
                                 attr_bygroup_x=attr_x2[,-1],
                                 attr_bygroup_y=attr_y2[m,-1])
        attr_comps <- attr_comps[which(!is.na(m)),]

    }
    colnames(attr_comps) <- gsub("attr", attribute, colnames(attr_comps))

    return(attr_comps)
}
#' calculate percentage of orf_b contained in orf_a
#' @param orf_a character string of ORF amino acid sequence
#' @param orf_b character string of ORF amino acid sequence
#' @return percentage of orf_b contained in orf_a
#' @export
#' @import stringdist
#' @examples
#' @author Beth Signal
orfSimilarity <- function(orf_a, orf_b){
    ((nchar(orf_a) + nchar(orf_b) - stringdist::stringdist(orf_a ,orf_b, method = "lcs"))/2) / nchar(orf_a)
}
