#' Filter out significant events from a whippet diff comparison
#' @param whippetDiff data.frame containing whippet diff data.
#' May be generated by readWhippetDIFFfiles()
#' @param probability minimum probability required to call event as significant
#' @param psiDelta minimum change in psi required to call an event as significant
#' @param eventTypes which event type to filter for? default = "all"
#' @param whippetCounts optional whippet counts data.frame for read count filtering.
#' May be generated by readWhippetPSIfiles()
#' @param minCounts minumum number of counts for all replicates
#' in at least one condition to call an event as significant
#' @param sampleTable data.frame with sample names and conditions.
#' Only needed if filtering with counts.
#' @return filtered whippet differential comparison data.frame
#' @export
#' @author Beth Signal
#' @examples
#' whippetFiles <- list.files(system.file("extdata","whippet/",
#' package = "GeneStructureTools"), full.names = TRUE)
#' diffFiles <- whippetFiles[grep(".diff", whippetFiles)]
#' whippetDiffSplice <- readWhippetDIFFfiles(diffFiles)
#' whippetDiffSplice.significant <- filterSignificantWhippetEvents(whippetDiffSplice)
filterSignificantWhippetEvents <- function(whippetDiff,
                                           probability = 0.95,
                                           psiDelta = 0.1,
                                           eventTypes = "all",
                                           whippetCounts = NULL,
                                           minCounts = NA,
                                           sampleTable){

    if(eventTypes == "all"){
        eventTypes <- unique(whippetDiff$type)
    }

    significantEvents <- whippetDiff[which(whippetDiff$probability > probability &
                                                 abs(whippetDiff$psi_delta) > psiDelta &
                                                 whippetDiff$type %in% eventTypes),]

    #filter by minimum read counts?
    if(!is.null(whippetCounts) & !is.na(minCounts)){
        m <- match(significantEvents$unique_name, whippetCounts$unique_name)
        n1 <- match(sampleTable$sample[sampleTable$condition %in%
                                                   unique(significantEvents$condition_1)],
                    colnames(whippetCounts))
        n2 <- match(sampleTable$sample[sampleTable$condition %in%
                                                   unique(significantEvents$condition_2)],
                    colnames(whippetCounts))
        significantEvents$condition_1_counts <- rowMeans(whippetCounts[m,n1])
        significantEvents$condition_2_counts <- rowMeans(whippetCounts[m,n2])

        keep <- which(apply(whippetCounts[m,n1], 1, function(x) all(x > minCounts)) |
                          apply(whippetCounts[m,n2], 1, function(x) all(x > minCounts)))

        significantEvents <- significantEvents[keep,]
    }

    return(significantEvents)

}
#' Compare open reading frames for two sets of paired transcripts
#' @param transcriptsX GRanges object with exon annotations for
#' all transcripts to be compared for the 'normal' condition
#' @param transcriptsY GRanges object with exon annotations for
#' all transcripts to be compared for the 'alternative' condition
#' @param BSgenome BSGenome object containing the genome for the species analysed
#' @param gtf.exons GRanges object made from a GTF containing exon coordinates
#' @param NMD Use NMD predictions? (Note: notNMD must be installed to use this feature)
#' @param orfPrediction What type of orf predictions to return. default: "allFrames"
#' @param compareBy compare isoforms by 'transcript' id, or aggregate all changes occuring by 'gene'
#' @param compareToGene compare alternative isoforms to all normal gene isoforms (in gtf.exons)
#' @param whippetEvents events generated by readWhippetDIFFfiles().
#' Use if PSI directionality should be taken into account when comparing isoforms.
#' @return Summarised ORF changes data.frame
#' @export
#' @import notNMD
#' @import GenomicRanges
#' @author Beth Signal
#' @examples
#' whippetFiles <- list.files(system.file("extdata","whippet/",
#' package = "GeneStructureTools"), full.names = TRUE)
#' diffFiles <- whippetFiles[grep(".diff", whippetFiles)]
#' whippetDiffSplice <- readWhippetDIFFfiles(diffFiles)
#' whippetCoords <- formatWhippetEvents(whippetDiffSplice)
#' gtf <- rtracklayer::import(system.file("extdata","example_gtf.gtf",
#' package = "GeneStructureTools"))
#' gtf.exons <- gtf[gtf$type=="exon"]
#' gtf.transcripts <- gtf[gtf$type=="transcript"]
#' g <- BSgenome.Mmusculus.UCSC.mm10::BSgenome.Mmusculus.UCSC.mm10
#' event.exonSkip <- whippetDiffSplice[which(whippetDiffSplice$type=="CE")[1],]
#' coords.exonSkip <- whippetCoords[whippetCoords$id %in% event.exonSkip$coord]
#' exons.exonSkip <- findExonContainingTranscripts(coords.exonSkip, gtf.exons,
#' variableWidth=0, findIntrons=FALSE, gtf.transcripts)
#' ExonSkippingTranscripts <- skipExonInTranscript(coords.exonSkip, exons.exonSkip, gtf.exons)
#' transcriptChangeSummary(ExonSkippingTranscripts[ExonSkippingTranscripts$set=="included_exon"],
#' ExonSkippingTranscripts[ExonSkippingTranscripts$set=="skipped_exon"],
#' BSgenome=g,gtf.exons)
transcriptChangeSummary <- function(transcriptsX,
                                    transcriptsY,
                                    BSgenome,
                                    gtf.exons,
                                    NMD = FALSE,
                                    compareBy="gene",
                                    orfPrediction = "allFrames",
                                    compareToGene = FALSE,
                                    whippetEvents = NULL){


    if(!is.null(whippetEvents)){
        allTranscripts <- c(transcriptsX, transcriptsY)
        type <- gsub("_X","",gsub("_Y","", allTranscripts$set))
        type <- gsub("skipped_exon", "CE", gsub("included_exon","CE", type))
        type <- gsub("retained_intron", "RI", gsub("spliced_intron","RI", type))

        m <- match(paste0(allTranscripts$whippet_id,"_",type), paste0(whippetEvents$coord,"_",
                                                                      whippetEvents$type))
        # A -- psi in condition 1 (A) is higher (i.e. included -- > skipped)
        normA <- which(whippetEvents$psi_a > whippetEvents$psi_b)
        # B -- psi in condition 2 (B) is higher (i.e. skipped -- > included)
        normB <- which(whippetEvents$psi_a < whippetEvents$psi_b)

        #sets for X (+A)
        setsX <- c(paste0(unique(type), "_Y"), "included_exon","retained_intron")
        #sets for Y (+B)
        setsY <- c(paste0(unique(type), "_X"), "skipped_exon","spliced_intron")

        transcriptsX <- allTranscripts[which((m %in% normA & allTranscripts$set %in% setsX) |
                                                 (m %in% normB & allTranscripts$set %in% setsY))]

        transcriptsY <- allTranscripts[which((m %in% normA & allTranscripts$set %in% setsY) |
                                                 (m %in% normB & allTranscripts$set %in% setsX))]

    }

    if(orfPrediction == "allFrames"){
        orfsX <- getOrfs(transcriptsX, BSgenome,returnLongestOnly = FALSE, allFrames = TRUE)
        orfsY <- getOrfs(transcriptsY, BSgenome,returnLongestOnly = FALSE, allFrames = TRUE)
    }else{
        orfsX <- getOrfs(transcriptsX, BSgenome,returnLongestOnly = TRUE)
        orfsY <- getOrfs(transcriptsY, BSgenome,returnLongestOnly = TRUE)
    }

    if(all(!grepl("[+]", orfsX$id))){

        if(orfPrediction == "allFrames"){
            Yid.withFrame <- paste0(unlist(lapply(
                str_split(orfsY$id, "[+]"),"[[",1)),"_", orfsY$frame)
            Xid.withFrame <- paste0(orfsX$id,"_", orfsX$frame)
            m <- match(Yid.withFrame, Xid.withFrame)
        }else{
            m <- match(unlist(lapply(str_split(orfsY$id, "[+]"),"[[",1)), orfsX$id)
        }

        orfsX<- orfsX[m,]
        orfsX$id <- orfsY$id
        #orfsX <- orfsX[which(!duplicated(orfsX$id)),]
    }

    # manual NMD

  if(NMD == TRUE){
      orfsX$nmd_prob <- notNMD::predictNMD(orfsX, "prob")
      orfsX$nmd_class <- notNMD::predictNMD(orfsX)
      orfsY$nmd_prob <- notNMD::predictNMD(orfsY, "prob")
      orfsY$nmd_class <- notNMD::predictNMD(orfsY)
  }

  orfsX <- orfsX[which(!is.na(orfsX$orf_length)),]
  orfsY <- orfsY[which(!is.na(orfsY$orf_length)),]

    if(compareToGene == TRUE){

      if(NMD == TRUE){
          orfAllGenes <- getOrfs(gtf.exons[gtf.exons$gene_id %in%
                                               unique(c(transcriptsX$gene_id,
                                                        transcriptsY$gene_id))],
                            BSgenome,returnLongestOnly = TRUE)
          orfAllGenes$nmd_prob <- notNMD::predictNMD(orfAllGenes, "prob")
          orfAllGenes$nmd_class <- notNMD::predictNMD(orfAllGenes)
          orfAllGenes <- orfAllGenes[orfAllGenes$nmd_prob]

      }else{
          orfAllGenes <- getOrfs(gtf.exons[gtf.exons$gene_id %in%
                                               unique(c(transcriptsX$gene_id,
                                                        transcriptsY$gene_id)) &
                                          gtf.exons$transcript_type=="protein_coding"],
                            BSgenome=BSgenome,returnLongestOnly = TRUE)
      }

      orfChange <- orfDiff(orfsX, orfsY, filterNMD = NMD, compareBy = "gene",
                           geneSimilarity = TRUE,allORFs = orfAllGenes,compareUTR = TRUE)
    }else{
        orfChange <- orfDiff(orfsX, orfsY, filterNMD = NMD, compareBy = "gene",
                           compareUTR = TRUE)
    }



    if(NMD == TRUE){
        orfsX$nmd_class_manual <- "nonsense_mediated_decay"
        orfsX$nmd_class_manual[orfsX$orf_length > 50 &
                                   (orfsX$min_dist_to_junction_b < 50 |
                                        orfsX$exon_b_from_final == 0)] <- "not_nmd"

        orfsX$nmd_prob_manual <- 1
        orfsX$nmd_prob_manual[orfsX$nmd_class_manual == "not_nmd"] <- 0

        orfsY$nmd_class_manual <- "nonsense_mediated_decay"
        orfsY$nmd_class_manual[orfsY$orf_length > 50 &
                                   (orfsY$min_dist_to_junction_b < 50 |
                                        orfsY$exon_b_from_final == 0)] <- "not_nmd"

        orfsY$nmd_prob_manual <- 1
        orfsY$nmd_prob_manual[orfsY$nmd_class_manual == "not_nmd"] <- 0

    }
if(NMD == TRUE){
  nmdChange <- attrChangeAltSpliced(orfsX,
                                    orfsY,
                                    attribute="nmd_prob",
                                    compareBy="gene",
                                    useMax=FALSE)
  m <- match(orfChange$id, nmdChange$id)
  orfChange <- cbind(orfChange, nmdChange[m,-1])

  nmdChangeMan <- attrChangeAltSpliced(orfsX,
                                    orfsY,
                                    attribute="nmd_prob_manual",
                                    compareBy="gene",
                                    useMax=FALSE)
  m <- match(orfChange$id, nmdChangeMan$id)
  orfChange <- cbind(orfChange, nmdChangeMan[m,-1])


}
  if(!is.null(whippetEvents)){
      m <- match(orfChange$id, whippetEvents$coord)
      orfChange <- cbind(whippetEvents[m,], orfChange)
  }
  return(orfChange)

}

#' Compare open reading frames for whippet differentially spliced events
#' @param significantEvents data.frame containing signficant whippet diff data.
#' @param whippetCoords whippet coordinates for signficant whippet diff data.
#' Generated by formatWhippetEvents().
#' @param eventTypes which event type to filter for? default = "all"
#' @param gtf.exons GRanges gtf annotation of exons
#' @param gtf.transcripts GRanges gtf annotation of transcripts
#' @param BSgenome BSGenome object containing the genome for the species analysed
#' @param jncCoords whippet junction locations. Only needed if event types include AA/AD/AF/AL.
#' @param NMD Use NMD predictions? (Note: notNMD must be installed to use this feature)
#' @return data.frame containing signficant whippet diff data and ORF change summaries
#' @export
#' @author Beth Signal
#' @examples
#' whippetFiles <- list.files(system.file("extdata","whippet/",
#' package = "GeneStructureTools"), full.names = TRUE)
#' diffFiles <- whippetFiles[grep(".diff", whippetFiles)]
#' whippetDiffSplice <- readWhippetDIFFfiles(diffFiles)
#' whippetCoords <- formatWhippetEvents(whippetDiffSplice)
#' jncFiles <- whippetFiles[grep(".jnc", whippetFiles)]
#' jncCoords <- readWhippetJNCfiles(jncFiles)
#' jncCoords <- formatJunctions(jncCoords)
#' gtf <- rtracklayer::import(system.file("extdata","example_gtf.gtf",
#' package = "GeneStructureTools"))
#' gtf.exons <- gtf[gtf$type=="exon"]
#' gtf.transcripts <- gtf[gtf$type=="transcript"]
#' g <- BSgenome.Mmusculus.UCSC.mm10::BSgenome.Mmusculus.UCSC.mm10
#' whippetTranscriptChangeSummary(significantEvents = whippetDiffSplice, whippetCoords,
#' eventTypes = "all",gtf.exons,gtf.transcripts,BSgenome = g,jncCoords = jncCoords, NMD=FALSE)
whippetTranscriptChangeSummary <- function(significantEvents,
                                           whippetCoords,
                                           eventTypes = "all",
                                           gtf.exons,
                                           gtf.transcripts,
                                           BSgenome,
                                           jncCoords,
                                           NMD = FALSE){

    if(eventTypes == "all"){
        eventTypes <- unique(significantEvents$type)
    }

    if(any(eventTypes == "CE") & any(significantEvents$type == "CE")){
        significantEvents.ce <- significantEvents[significantEvents$type=="CE",]

        # get whippet exon coordinates
        ranges.ce <- whippetCoords[whippetCoords$id %in% significantEvents.ce$coord]
        # find exons in the gtf that overlap whippet exons
        exons.ce <- findExonContainingTranscripts(ranges.ce, gtf.exons,
                                                  variableWidth=0, findIntrons=FALSE,
                                                  gtf.transcripts)

        # make skipped exon transcripts
        skippedExonTranscripts <- skipExonInTranscript(ranges.ce, exons.ce,
                                                         gtf.exons, glueExons = TRUE)

        orfChanges.ce <- transcriptChangeSummary(
            skippedExonTranscripts[skippedExonTranscripts$set=="included_exon"],
            skippedExonTranscripts[skippedExonTranscripts$set=="skipped_exon"],
            BSgenome = BSgenome,NMD = NMD, whippetEvents=significantEvents.ce)
        # add to significantEvents.ce
        m <- match(significantEvents.ce$coord, orfChanges.ce$id)
        significantEvents.ce <- cbind(significantEvents.ce, orfChanges.ce[m,-1])

        if(exists("SignificantEvents.withORF")){
            SignificantEvents.withORF <- rbind(SignificantEvents.withORF,
                                               significantEvents.ce)
        }else{
            SignificantEvents.withORF <- significantEvents.ce
        }
    }
    if(any(eventTypes == "RI") & any(significantEvents$type == "RI")){
        significantEvents.ri <- significantEvents[significantEvents$type=="RI",]

        # get whippet exon coordinates
        ranges.ri <- whippetCoords[whippetCoords$id %in% significantEvents.ri$coord]
        # need to extend for whippet coords
        start(ranges.ri) <- start(ranges.ri) -1
        end(ranges.ri) <- end(ranges.ri) +1
        # find introns in the gtf that overlap whippet introns
        exons.ri <- findIntronContainingTranscripts(ranges.ri, gtf.exons)
        # add the intron into transcripts
        retainedIntronTranscripts <- addIntronInTranscript(ranges.ri, exons.ri,
                                                           gtf.exons, glueExons = TRUE)

        orfChanges.ri <- transcriptChangeSummary(
            retainedIntronTranscripts[retainedIntronTranscripts$set=="spliced_intron"],
            retainedIntronTranscripts[retainedIntronTranscripts$set=="retained_intron"],
            BSgenome = BSgenome,NMD = NMD, whippetEvents=significantEvents.ri)
        # add to significantEvents.ce
        m <- match(significantEvents.ri$coord, orfChanges.ri$id)
        significantEvents.ri <- cbind(significantEvents.ri, orfChanges.ri[m,-1])

        if(exists("SignificantEvents.withORF")){
            SignificantEvents.withORF <- rbind(SignificantEvents.withORF,
                                               significantEvents.ri)
        }else{
            SignificantEvents.withORF <- significantEvents.ri
        }

    }
    if(any(eventTypes %in% c("AA","AD","AF","AL")) &
       any(significantEvents$type %in% c("AA","AD","AF","AL"))){

        events.junctions <- eventTypes[eventTypes %in% c("AA","AD","AF","AL")]
        events.significant <- unique(significantEvents$type)
        events.significant <- events.significant[events.significant %in% events.junctions]

        for(e in seq_along(events.significant)){

            event <- events.significant[e]
            significantEvents.jnc <- significantEvents[significantEvents$type==event,]

            # get whippet exon coordinates
            ranges.jnc <- whippetCoords[whippetCoords$id %in% significantEvents.jnc$coord]

            junctionPairs <- findJunctionPairs(ranges.jnc, jncCoords, type=event)

            # check for pairs
            ids.x <- unique(junctionPairs$whippet_id[junctionPairs$set=="X"])
            ids.x <- ids.x[ids.x %in% unique(junctionPairs$whippet_id[junctionPairs$set=="Y"])]

            significantEvents.jnc <-significantEvents.jnc[which(significantEvents.jnc$coord %in%
                                                                    ids.x),]
            ranges.jnc <- ranges.jnc[which(ranges.jnc$id %in% ids.x),]
            junctionPairs <- junctionPairs[which(junctionPairs$whippet_id %in% ids.x),]

            if(nrow(significantEvents.jnc) > 0){
              # make transcripts with alternative junction usage
              altTranscripts <- replaceJunction(junctionPairs, ranges.jnc, gtf.exons, type=event)
              orfChanges.jnc <- transcriptChangeSummary(
                  transcriptsX = altTranscripts[altTranscripts$set==paste0(event, "_X")],
                  transcriptsY = altTranscripts[altTranscripts$set==paste0(event, "_Y")],
                  BSgenome = BSgenome,NMD = NMD, whippetEvents=significantEvents.jnc)

              # add to significantEvents
              m <- match(significantEvents.jnc$coord, orfChanges.jnc$id)
              significantEvents.jnc <- cbind(significantEvents.jnc, orfChanges.jnc[m,-1])

              if(exists("SignificantEvents.withORF")){
                  SignificantEvents.withORF <- rbind(SignificantEvents.withORF,
                                                     significantEvents.jnc)
              }else{
                  SignificantEvents.withORF <- significantEvents.jnc
              }
            }
        }

    }

    return(SignificantEvents.withORF)
}
#' Compare open reading frames for whippet differentially spliced events
#' @param significantEvents  data.frame containing information from the
#' per_intron_results.tab file output from leafcutter.
#' @param combineGeneEvents combine clusters occuring in the same gene?
#' Currently NOT FUNCTIONAL
#' @param gtf.exons GRanges gtf annotation of exons
#' @param BSgenome BSGenome object containing the genome for the species analysed
#' @param NMD Use NMD predictions? (Note: notNMD must be installed to use this feature)
#' @param showProgressBar show a progress bar of alternative isoform generation?
#' @return data.frame containing signficant whippet diff data and ORF change summaries
#' @export
#' @importFrom utils setTxtProgressBar
#' @importFrom utils txtProgressBar
#' @importFrom stringr str_split
#' @import GenomicRanges
#' @author Beth Signal
#' @examples
#' leafcutterFiles <- list.files(system.file("extdata","leafcutter/",
#' package = "GeneStructureTools"), full.names = TRUE)
#' leafcutterIntrons <- read.delim(leafcutterFiles[
#' grep("intron_results", leafcutterFiles)],stringsAsFactors=FALSE)
#' gtf <- rtracklayer::import(system.file("extdata","example_gtf.gtf",
#' package = "GeneStructureTools"))
#' gtf.exons <- gtf[gtf$type=="exon"]
#' g <- BSgenome.Mmusculus.UCSC.mm10::BSgenome.Mmusculus.UCSC.mm10
#' leafcutterTranscriptChangeSummary(significantEvents = leafcutterIntrons,
#' gtf.exons=gtf.exons,BSgenome = g,NMD=FALSE)

leafcutterTranscriptChangeSummary <- function(significantEvents,
                                              combineGeneEvents=FALSE,
                                              gtf.exons,
                                              BSgenome,
                                              NMD = FALSE,
                                              showProgressBar=TRUE){


    geneEvents <- as.data.frame(table(significantEvents$ensemblID,
                                      significantEvents$clusterID))
    geneEvents <- geneEvents[geneEvents$Freq!=0,]

    if(combineGeneEvents == FALSE){
        if(showProgressBar){
            message(paste0("Generating alternative isoforms for ",
                           nrow(geneEvents), " clusters:"))
            pb <- utils::txtProgressBar(min = 0, max = nrow(geneEvents), style = 3)
        }

        altIso <- alternativeIntronUsage(significantEvents[
            significantEvents$clusterID == geneEvents$Var2[1],],
            gtf.exons)
        if(showProgressBar){utils::setTxtProgressBar(pb, 1)}

        if(nrow(geneEvents) > 1){
            for(i in 2:nrow(geneEvents)){
                altIntronLocs = significantEvents[
                    significantEvents$clusterID == geneEvents$Var2[i],]
                altIntronLocs <- altIntronLocs[altIntronLocs$verdict=="annotated",]
                if(nrow(altIntronLocs) > 1){
                    altIso1 <- alternativeIntronUsage(altIntronLocs, gtf.exons)
                    altIso <- c(altIso, altIso1)
                }
                if(showProgressBar){utils::setTxtProgressBar(pb, i)}

            }
        }
    }else{
        genes <- unique(geneEvents$Var1)
        if(showProgressBar){
            message(paste0("Generating alternative isoforms for ",
                           nrow(geneEvents), " genes:"))
            pb <- utils::txtProgressBar(min = 0, max = length(genes), style = 3)
        }
        for(j in seq_along(genes)){
            clusters <- geneEvents[geneEvents$Var1==genes[j],]

            altIso <- alternativeIntronUsage(significantEvents[
                significantEvents$clusterID == clusters$Var2[1],],
                gtf.exons)

            if(nrow(clusters) > 1){
                for(i in 2:nrow(clusters)){
                    altIntronLocs = significantEvents[
                        significantEvents$clusterID == clusters$Var2[i],]
                    altIntronLocs <- altIntronLocs[altIntronLocs$verdict=="annotated",]
                    if(nrow(altIntronLocs) > 1){
                        altIso1 <- alternativeIntronUsage(altIntronLocs, c(gtf.exons, altIso))
                        altIso <- c(altIso, altIso1)
                    }

                }
            }
            if(showProgressBar){utils::setTxtProgressBar(pb, j)}

        }


    }
    altIso$spliced_id <- unlist(lapply(
        stringr::str_split(altIso$transcript_id, " "),"[[",2))

    transcriptsX <- altIso[grep("dnre", altIso$transcript_id)]
    transcriptsY <- altIso[grep("upre", altIso$transcript_id)]

    orfDiff <- transcriptChangeSummary(transcriptsX,
                                       transcriptsY,
                                       BSgenome = BSgenome,
                                       NMD = NMD)
    m <- match(gsub("_","",significantEvents$clusterID), orfDiff$id)
    significantEvents.withORF <- cbind(significantEvents, orfDiff[m,-1])
    #significantEvents.withORF <- significantEvents.withORF[!duplicated(m),]

    return(significantEvents.withORF)
}
