#' Read in a list of whippet .jnc.gz files and format as a data.frame
#' @param files vector of *.jnc.gz file names
#' @return data.frame with junction counts for all files
#' @export
#' @import data.table
#' @examples
#' @author Beth Signal
readWhippetJNCfiles <- function(files){

    for(f in seq_along(files)){
        whip <- data.table::fread(paste0("zcat < ", files[f]), data.table=F)
        colnames(whip) <- c("chromosome","start","end","id","count","strand")

        if(exists("whip.all")){
            m <- match(whip$id, whip.all$id)
            n <- match(whip.all$id, whip$id)
            whip.all <- cbind(whip.all, new_count=whip[n,5])
            add_cols <- ncol(whip.all) - 6

            if(any(is.na(m))){
                whip.add <- whip[which(is.na(m)),c(1:4,6)]
                for(c in 1:(add_cols)){
                    whip.add <- cbind(whip.add, count=NA)
                }
                whip.add <- cbind(whip.add, count=whip$count[which(is.na(m))])
                colnames(whip.add) <- colnames(whip.all)
                whip.all <- rbind(whip.all,whip.add)
            }
        }else{
            whip.all <- whip[,c(1:4,6,5)]
        }
    }

    colnames(whip.all)[-(1:5)] <- gsub(".jnc.gz","", basename(files))
    return(whip.all)
}

#' Read in a list of whippet .psi.gz files and format as a data.frame
#' @param files vector of *.psi.gz file names
#' @return data.frame with junction counts for all files
#' @export
#' @import data.table
#' @examples
#' @author Beth Signal
readWhippetPSIfiles <- function(files, attribute="Total_Reads", maxNA=NA){

    for(f in seq_along(files)){
        whip <- data.table::fread(paste0("zcat < ", files[f]), data.table=F)
        wanted_col <- which(colnames(whip) == attribute)

        if(exists("whip.all")){
            whip.all <- cbind(whip.all, whip[,wanted_col])
        } else{
            whip.all <- whip[,c(1:5,wanted_col)]
        }
    }

    colnames(whip.all)[-(1:5)] <- gsub(".psi.gz","", basename(files))
    whip.all$na_count <- apply(whip.all[,-c(1:5)], 1, function(x) length(which(is.na(x))))
    whip.all$unique_name <- with(whip.all, paste0(Gene,"_",Coord,"_",Type,"_",Node))
    if(!is.na(maxNA)){
        keep <- which(whip.all$na_count <= maxNA)
        whip.all <- whip.all[keep,]
    }
    return(whip.all)
}

#' Read in a list of whippet .diff.gz files and format as a data.frame
#' @param files vector of *.diff.gz file names
#' @return data.frame with junction counts for all files
#' @export
#' @import data.table
#' @examples
#' @author Beth Signal
readWhippetDIFFfiles <- function(files){

    for(f in seq_along(files)){
        whip <- data.table::fread(paste0("zcat < ", files[f]), data.table=F, skip=1)

        #remove the NA column
        keepcol <- which(apply(whip, 2, function(x) all(!is.na(x))))
        whip <- whip[,keepcol]

        colnames(whip) <- c("gene", "node","coord","strand","type",
                                 "psi_a","psi_b","psi_delta",
                                 "probability","complexity","entropy")
        whip$unique_name <- with(whip, paste0(gene,"_",coord,"_",type,"_",node))
        whip$comparison <- gsub(".diff.gz","", basename(files[f]))

        if(exists("whip.all")){
            whip.all <- rbind(whip.all, whip)
        } else{
            whip.all <- whip
        }
    }
    return(whip.all)
}

#' Format Whippet co-ordinates as a GRanges object
#' @param whippet data.frame containing event location information. May be generated by readWhippetDIFFfiles()
#' @return GRanges object with events
#' @export
#' @import GenomicRanges
#' @import IRanges
#' @import stringr
#' @examples
#' @author Beth Signal
formatWhippetEvents <- function(whippet){

    whippet <- whippet[!duplicated(whippet$coord),]

    chromosome <- unlist(lapply(str_split(whippet$coord,":"), "[[",1))
    range <- unlist(lapply(str_split(whippet$coord,":"), "[[",2))
    start <- unlist(lapply(str_split(range,"-"), "[[",1))
    end <- unlist(lapply(str_split(range,"-"), "[[",2))

    event_ranges <- GRanges(seqnames=Rle(chromosome),
                          ranges=IRanges(start=as.numeric(start),
                                         end=as.numeric(end)),
                          strand=whippet$strand, id=whippet$coord)
    return(event_ranges)
}

